File: 1601-1700/1681-1690/1684E_Count_the_Number_of_Consistent_Strings/kt/bit.kt
// 1684. Count the Number of Consistent Strings
// https://leetcode.com/problems/count-the-number-of-consistent-strings/description/
// T.C.: O(m+n*k)
// S.C.: O(1)
class Solution {
  fun countConsistentStrings(allowed: String, words: Array<String>): Int {
    var count = 0
    var allowedBit = 0
    for (char in allowed) {
      allowedBit = allowedBit or (1 shl (char - 'a'))
    }

    for (word in words) {
      var isConsistent = true
      for (char in word) {
        if ((allowedBit shr (char - 'a')) and 1 == 0) {
          isConsistent = false
          break
        }
      }
      if (isConsistent) {
        count++
      }
    }
    return count
  }
}

fun main() {
  var allowed = "ab"
  var words = arrayOf("ad", "bd", "aaab", "baa", "badab")
  var expected = 2
  var result = Solution().countConsistentStrings(allowed, words)
  println("$result -> ${result == expected}")

  allowed = "abc"
  words = arrayOf("a", "b", "c", "ab", "ac", "bc", "abc")
  expected = 7
  result = Solution().countConsistentStrings(allowed, words)
  println("$result -> ${result == expected}")

  allowed = "cad"
  words = arrayOf("cc", "acd", "b", "ba", "bac", "bad", "ac", "d")
  expected = 4
  result = Solution().countConsistentStrings(allowed, words)
  println("$result -> ${result == expected}")
}

File: 1601-1700/1681-1690/1684E_Count_the_Number_of_Consistent_Strings/bit.js
// 1684. Count the Number of Consistent Strings
// https://leetcode.com/problems/count-the-number-of-consistent-strings/description/
// T.C.: O(m+n*k)
// S.C.: O(1)
/**
 * @param {string} allowed
 * @param {string[]} words
 * @return {number}
 */
var countConsistentStrings = function (allowed, words) {
  let count = 0;
  let allowedBit = 0;
  const a = 'a'.charCodeAt();
  for (let i = 0; i < allowed.length; i++) {
    allowedBit |= 1 << (allowed.charCodeAt(i) - a);
  }

  for (const word of words) {
    let isConsistent = true;
    for (let i = 0; i < word.length; i++) {
      const index = word.charCodeAt(i) - a;
      if (((1 << index) & allowedBit) === 0) {
        isConsistent = false;
        break;
      }
    }
    if (isConsistent) {
      count++;
    }
  }
  return count;
};

var allowed = 'ab',
  words = ['ad', 'bd', 'aaab', 'baa', 'badab'];
var expected = 2;
var result = countConsistentStrings(allowed, words);
console.log(result, result === expected);

var allowed = 'abc',
  words = ['a', 'b', 'c', 'ab', 'ac', 'bc', 'abc'];
var expected = 7;
var result = countConsistentStrings(allowed, words);
console.log(result, result === expected);

var allowed = 'cad',
  words = ['cc', 'acd', 'b', 'ba', 'bac', 'bad', 'ac', 'd'];
var expected = 4;
var result = countConsistentStrings(allowed, words);
console.log(result, result === expected);
